src/protocol/sort/bit_permutation.rs

```diff
@@ -16,13 +21,17 @@ use futures::future::try_join_all;
 /// by K. Chida, K. Hamada, D. Ikarashi, R. Kikuchi, N. Kiribuchi, and B. Pinkas
 /// <https://eprint.iacr.org/2019/695.pdf>.
 #[derive(Debug)]
-pub struct BitPermutation<'a, F> {
-    input: &'a [Replicated<F>],
+pub struct BitPermutation<'a, F, S> {
+    input: &'a [S],
+    _marker: PhantomData<F>,
 }
 
-impl<'a, F: Field> BitPermutation<'a, F> {
-    pub fn new(input: &'a [Replicated<F>]) -> BitPermutation<'a, F> {
-        Self { input }
+impl<'a, S: SecretSharing<F> + Copy, F: Field> BitPermutation<'a, F, S> {
```

Comment:
Why did you need to add `+ Copy`? I thought we were trying to avoid that one. 

I see that you're introducing a trait bound that `S: SecretSharing<F>`, and `F: Field`. That's good. We don't need these to be totally generic as they appear in the struct definition. I think if this protocol was just a single public async function these bounds would be enough to eliminate the need for that `PhantomData` stuff.

Reply:
Copy is needed since we duplicate the input at ".chain(input.iter().copied())". Is there a reason why Copy in this scenario would be harmful?

Reply:
You can use .cloned() instead and rely on the Clone implementation.